### PR TITLE
NIP101 censorship resistant marketplace

### PR DESCRIPTION
**Pull request by arcbtc**
Some Nostr note kinds for a censorship resistant marketplace. 

This is based on Diagon Alley, the protocol that partly influenced Nostr, where clients are market stalls of products, and relays are online product indexers. 
https://github.com/lnbits/Diagon-Alley

The best way to use NIP101 is if the buyer and merchant are both clients, an ecommerce management client for the merchant, and a shop client for the buyer (Nostr clients could be run as a service, or by the users themselves).   

The ordering process is end-to-end encrypted between buyer and merchant.

### CODE REVIEW

**Comment by fiatjaf** [comment]
Closed by accident.

**Comment by fiatjaf** [comment]
I have many complaints:

I don't understand why we need `encrypted_payment` (or what would be go there, an URL? Why can't we have the URL in the product announcement directly?).

I also don't understand why we need `encrypted_receipt`. I thought this was supposed to work with Lightning payments. In this case the receipt is there already on Lightning.

I also think we need a single `shop_announcement` containing all the products of the shop instead of different events for each product, because then clients and relays can just delete the previous events and keep the latest one as the single source of truth. `encrypted_shipping` could be included in the `shop_announcement`, maybe as a set of rules for shipping conditions and prices, but I'm not sure.

`feedback` in my opinion is also better done as a simple `text_note` referencing the `shop_announcement`.

So in my opinion we should only have two types of new messages here.

And finally this should be NIP 5, not 101, even if it defines event kinds 101 and 102. But maybe it should define kinds 5 and 6 also. I don't know anymore what rule is good to keep for these numbers.

**Comment by arcbtc** [comment]
For gathering feedback, having a kind would make it easier for shop clients. The kind could also be used by other nostr use cases, perhaps for support tickets, etc

This is not LN specific, thats an assumption not mentioned in the NIP, you can have conventional payments using these kinds

I will have a look at tightening NIP101 somehow

File: nips/101.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+A suite of 6 `kinds` to build a censorship resistant marketplace on Nostr network
+
+* `kind` `101` `product`
+* `kind` `102` `encrypted_order`
+* `kind` `103` `encrypted_payment`
+* `kind` `104` `encrypted_receipt`
+* `kind` `105` `encrypted_shipping`
+* `kind` `106` `feedback`
+
+End-to-end encryption between buyer and merchant is achieved using NIP04 encryption scheme https://github.com/fiatjaf/nostr/pull/10
```
Comment (MaxHillebrand) [inline nips/101.md:18]:
```suggestion
End-to-end encryption between buyer and merchant is achieved using NIP04 encryption scheme https://github.com/fiatjaf/nostr/blob/master/nips/04.md
```

File: nips/101.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        ... // if p tag of the merchants public-key is present, product should be removed
+      ]
+      content: <stringified JSON object>,
+      sig: <64-bytes signature of the sha256 hash of the serialized event data, which is the same as the "id" field>,
+    }
+
+
+## `kind` `102` `encrypted_order` (buyer -> relay -> merchant)
+
+`102`: `encrypted_order`: the `content` is set to a base-64 encoded, aes-256-cbc encrypted, stringified JSON object `{quantity: <integer>, name: <string>, address: <string>, phone: <string>, email: <string>}` passing order details to merchant.
```
Comment (MaxHillebrand) [inline nips/101.md:53]: Should this include `nostr pubkey`?

File: nips/101.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        ["e",  <32-bytes hex of the product id>, <recommended relay URL>],
+        ["s", <32-byte sha256 hash of the shared cipher>],
+      ]
+      content: <base-64 encoded, aes-256-cbc encrypted, stringified JSON object>,
+      sig: <64-bytes signature of the sha256 hash of the serialized event data, which is the same as the "id" field>,
+    }
+
+## `kind` `103` `encrypted_payment` (merchant -> relay -> buyer)
+
+`103`: `encrypted_payment`: the `content` is set to a base-64 encoded, aes-256-cbc encrypted, stringified JSON object `{data: <string>, url: <string>}` passing payment details to the buyer.
```
Comment (MaxHillebrand) [inline nips/101.md:74]: what does `url` do?

**Comment by MaxHillebrand** [comment]
Do I understand correctly, that this does not provide buyer privacy, in the sense that the relay knows that a certain pubkey is talking to the selling pubkey?

**Pull request closed without merging**